File: src/darmoshark/tuning_packet.py
"""Builders for the sensor and click tuning packets."""

from darmoshark.dms_commands import DmsCommands
from darmoshark.protocol import DarmosharkProtocol


class TuningPacket:
    """Debounce, lift-off distance, scroll and sleep timer packets.

    All of them ride the 20-byte short report (0xB5) and are acknowledged with
    an 0xE4 frame echoing the opcode.
    """

    debounceMinimum = 0
    debounceMaximum = 20
    liftOffValues = (1, 2)          # M3 profile declares only low/high
    sleepMinimumMinutes = 0
    sleepMaximumMinutes = 255

    @staticmethod
    def buildDebounce(milliseconds):
        if not TuningPacket.debounceMinimum <= milliseconds <= TuningPacket.debounceMaximum:
            raise ValueError(
                f"debounce must be {TuningPacket.debounceMinimum}-"
                f"{TuningPacket.debounceMaximum} ms, got {milliseconds}")
        payload = bytearray(DarmosharkProtocol.shortPayloadSize)
        payload[0] = DmsCommands.setButtonDebounce
        payload[1] = milliseconds
        return DarmosharkProtocol.shortReportId, bytes(payload)

    @staticmethod
    def buildSensor(liftOff=1, wave=1, line=2, motion=1, scroll=1, eSports=1):
        """Sensor block: lift-off distance plus the assorted on/off toggles."""
        if liftOff not in TuningPacket.liftOffValues:
            raise ValueError(
                f"liftOff must be one of {TuningPacket.liftOffValues}, got {liftOff}")
        payload = bytearray(DarmosharkProtocol.shortPayloadSize)
        payload[0] = DmsCommands.setSensorLiftCutoff
        payload[1] = liftOff
        payload[2] = wave
        payload[3] = line
        payload[4] = motion
        payload[6] = scroll
        payload[7] = eSports
        return DarmosharkProtocol.shortReportId, bytes(payload)

    @staticmethod
    def buildScroll(speed, inertia, spl):
        for name, value in (("speed", speed), ("inertia", inertia), ("spl", spl)):
            if not 0 <= value <= 0xFF:
                raise ValueError(f"{name} must fit in a byte, got {value}")
        payload = bytearray(DarmosharkProtocol.shortPayloadSize)
        payload[0] = DmsCommands.setScroll
        payload[1] = speed
        payload[2] = inertia
        payload[3] = spl
        return DarmosharkProtocol.shortReportId, bytes(payload)

    @staticmethod
    def buildSleep(minutes, mode="set"):
        modes = {"set": 1, "get": 2}
        if mode not in modes:
            raise ValueError(f"mode must be 'set' or 'get', got {mode!r}")
        if not TuningPacket.sleepMinimumMinutes <= minutes <= TuningPacket.sleepMaximumMinutes:
            raise ValueError(
                f"sleep must be {TuningPacket.sleepMinimumMinutes}-"
                f"{TuningPacket.sleepMaximumMinutes} minutes, got {minutes}")
        payload = bytearray(DarmosharkProtocol.shortPayloadSize)
        payload[0] = DmsCommands.deviceTime
        payload[1] = modes[mode]
        payload[2] = minutes
        return DarmosharkProtocol.shortReportId, bytes(payload)

File: crates/darmoshark/src/packets/tuning_packet.rs
use crate::error::darmoshark_error::{DarmosharkError, DarmosharkResult};
use crate::packets::packet::Packet;
use crate::packets::sleep_mode::SleepMode;
use crate::protocol::darmoshark_protocol::DarmosharkProtocol;
use crate::protocol::dms_commands::DmsCommands;

/// Debounce, lift-off distance, scroll and sleep timer packets.
///
/// All of them ride the 20-byte short report (0xB5) and are acknowledged with
/// an 0xE4 frame echoing the opcode.
pub struct TuningPacket;

impl TuningPacket {
  pub const debounceMinimum: u8 = 0;
  pub const debounceMaximum: u8 = 20;
  pub const liftOffValues: [u8; 2] = [1, 2]; // M3 profile declares only low/high
  pub const sleepMinimumMinutes: u8 = 0;
  pub const sleepMaximumMinutes: u8 = 255;

  pub fn build_debounce(milliseconds: u32) -> DarmosharkResult<Packet> {
    let range = u32::from(Self::debounceMinimum)..=u32::from(Self::debounceMaximum);
    if !range.contains(&milliseconds) {
      return Err(DarmosharkError::Invalid(format!(
        "debounce must be {}-{} ms, got {milliseconds}",
        Self::debounceMinimum,
        Self::debounceMaximum
      )));
    }
    let mut payload = vec![0u8; DarmosharkProtocol::shortPayloadSize];
    payload[0] = DmsCommands::setButtonDebounce;
    payload[1] = milliseconds as u8;
    Ok(Packet::new(DarmosharkProtocol::shortReportId, payload))
  }

  /// Sensor block: lift-off distance plus the assorted on/off toggles.
  ///
  /// The vendor defaults are `wave 1, line 2, motion 1, scroll 1, eSports 1`.
  pub fn build_sensor(
    lift_off: u8,
    wave: u8,
    line: u8,
    motion: u8,
    scroll: u8,
    e_sports: u8,
  ) -> DarmosharkResult<Packet> {
    if !Self::liftOffValues.contains(&lift_off) {
      return Err(DarmosharkError::Invalid(format!(
        "liftOff must be one of {:?}, got {lift_off}",
        Self::liftOffValues
      )));
    }
    let mut payload = vec![0u8; DarmosharkProtocol::shortPayloadSize];
    payload[0] = DmsCommands::setSensorLiftCutoff;
    payload[1] = lift_off;
    payload[2] = wave;
    payload[3] = line;
    payload[4] = motion;
    payload[6] = scroll;
    payload[7] = e_sports;
    Ok(Packet::new(DarmosharkProtocol::shortReportId, payload))
  }

  pub fn build_scroll(speed: u8, inertia: u8, spl: u8) -> Packet {
    let mut payload = vec![0u8; DarmosharkProtocol::shortPayloadSize];
    payload[0] = DmsCommands::setScroll;
    payload[1] = speed;
    payload[2] = inertia;
    payload[3] = spl;
    Packet::new(DarmosharkProtocol::shortReportId, payload)
  }

  pub fn build_sleep(minutes: u32, mode: SleepMode) -> DarmosharkResult<Packet> {
    let range = u32::from(Self::sleepMinimumMinutes)..=u32::from(Self::sleepMaximumMinutes);
    if !range.contains(&minutes) {
      return Err(DarmosharkError::Invalid(format!(
        "sleep must be {}-{} minutes, got {minutes}",
        Self::sleepMinimumMinutes,
        Self::sleepMaximumMinutes
      )));
    }
    let mut payload = vec![0u8; DarmosharkProtocol::shortPayloadSize];
    payload[0] = DmsCommands::deviceTime;
    payload[1] = mode.marker();
    payload[2] = minutes as u8;
    Ok(Packet::new(DarmosharkProtocol::shortReportId, payload))
  }
}
